zk/stages/stage_sequence_execute.go

```diff
@@ -79,29 +75,69 @@ func SpawnSequencingStage(
 		if err = cfg.datastreamServer.WriteWholeBatchToStream(logPrefix, sdb.tx, sdb.hermezDb.HermezDbReader, lastBatch, injectedBatchBatchNumber); err != nil {
 			return err
 		}
+		if err = stages.SaveStageProgress(sdb.tx, stages.DataStream, 1); err != nil {
+			return err
+		}
 
 		return sdb.tx.Commit()
 	}
 
-	tryHaltSequencer(batchContext, batchState.batchNumber)
-
-	if err := utils.UpdateZkEVMBlockCfg(cfg.chainConfig, sdb.hermezDb, logPrefix); err != nil {
-		return err
+	if batchState.hasExecutorForThisBatch {
+		// identify a stream gap i.e. a sequencer restart without an ack back from an executor.
+		// in this case we need to unwind the state so that we match the datastream height
+		streamProgress, err := stages.GetStageProgress(sdb.tx, stages.DataStream)
+		if err != nil {
+			return err
+		}
+		if streamProgress > 0 && streamProgress < executionAt {
+			block, err := rawdb.ReadBlockByNumber(sdb.tx, streamProgress)
+			if err != nil {
+				return err
+			}
+			log.Warn(fmt.Sprintf("[%s] Unwinding due to a datastream gap", logPrefix),
+				"streamHeight", streamProgress,
+				"sequencerHeight", executionAt,
+			)
+			u.UnwindTo(streamProgress, block.Hash())
+			return nil
+		}
 	}
 
-	batchCounters, err := prepareBatchCounters(batchContext, batchState, isLastBatchPariallyProcessed)
+	lastBatchSealed, err := checkIfLastBatchIsSealed(batchContext)
 	if err != nil {
 		return err
 	}
 
-	if !isLastBatchPariallyProcessed {
-		// handle case where batch wasn't closed properly
-		// close it before starting a new one
-		// this occurs when sequencer was switched from syncer or sequencer datastream files were deleted
-		// and datastream was regenerated
-		if err = finalizeLastBatchInDatastreamIfNotFinalized(batchContext, batchState, executionAt); err != nil {
+	if !lastBatchSealed {
```

Comment:
Can we just use "runBlocksLoops = false" in this case and let the rest of the code execute instead of creating a "special" case only for closing a batch. The goal is to eliminate the duplicate code.

Reply:
That sounds like a good plan!  Nice idea

Reply:
I made this change but it gets confusing, and for the code we drop we need to populate more in it's place.  Skipping the block loop causes the new batch to be sealed, but we want to seal the last batch instead.  I think it's more readable to have this self contained bit of code that is very explicit in what it's doing

Reply:
Is it worth making final try - decrement the batchNumber in batchState and set runBlockLoop = false. It should work. If not - leave it as it is.